src/Stratis.Bitcoin.Features.SignalR/Broadcasters/BroadcasterBase.cs

```diff
@@ -36,16 +39,52 @@ public void Init(ClientEventBroadcasterSettings broadcasterSettings)
                 $"Broadcast {this.GetType().Name}",
                 async token =>
                 {
-                    foreach (IClientEvent clientEvent in this.GetMessages())
+                    using (var cancellationTokenSource = new CancellationTokenSource())
                     {
-                        await this.eventsHub.SendToClientsAsync(clientEvent)
-                            .ConfigureAwait(false);
+                        cancellationTokenSource.CancelAfter(new TimeSpan(0, 0, 10));
+
+                        using (var linkedTokenSource =
+                            CancellationTokenSource.CreateLinkedTokenSource(cancellationTokenSource.Token, token))
+                        {
+                            try
+                            {
+                                var messages = await this.GetMessages(linkedTokenSource.Token);
+                                foreach (IClientEvent clientEvent in messages)
+                                {
+                                    await this.eventsHub.SendToClientsAsync(clientEvent);
+                                }
+                            }
+                            catch (Exception ex)
+                            {
+                                this.logger.LogError($"{this.GetType().Name} Error in GetMessages",
+                                    ex);
+                            }
+                        }
                     }
                 },
                 this.nodeLifetime.ApplicationStopping,
-                repeatEvery: TimeSpan.FromSeconds(Math.Max(broadcasterSettings.BroadcastFrequencySeconds, 5)));
+                repeatEvery:
+                TimeSpan.FromSeconds(Math.Max(broadcasterSettings.BroadcastFrequencySeconds, 5)));
+
+            this.OnInitialise();
+        }
+
+        protected abstract Task<IEnumerable<IClientEvent>> GetMessages(CancellationToken cancellationToken);
+
+        protected virtual void Dispose(bool disposing)
```

Comment:
Do we need to implement this dispose pattern? Can't we just call `Dispose` and then dispose the `asyncloop` in there?

Reply:
Yes i think so as the class is abstract and inherited. This is actually used on the `WalletInfoBroadcaster` to unsubscribe from incoming signalR messages